pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -435,59 +540,90 @@ func (e *TokensController) removeSecretReferenceIfNeeded(serviceAccount *api.Ser
 		}
 	}
 	serviceAccount.Secrets = secrets
-
 	_, err = serviceAccounts.Update(serviceAccount)
-	if err != nil {
-		return err
+	// Ignore NotFound errors when attempting to remove a reference
+	if apierrors.IsNotFound(err) {
+		return nil
 	}
-
-	return nil
+	return err
 }
 
-// getServiceAccount returns the ServiceAccount referenced by the given secret. If the secret is not
-// of type ServiceAccountToken, or if the referenced ServiceAccount does not exist, nil is returned
-func (e *TokensController) getServiceAccount(secret *api.Secret, fetchOnCacheMiss bool) (*api.ServiceAccount, error) {
-	name, _ := serviceAccountNameAndUID(secret)
-	if len(name) == 0 {
+func (e *TokensController) getServiceAccount(ns string, name string, uid types.UID, fetchOnCacheMiss bool) (*api.ServiceAccount, error) {
+	// Look up in cache
+	obj, exists, err := e.serviceAccounts.GetByKey(ns + "/" + name)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		sa, ok := obj.(*api.ServiceAccount)
+		if !ok {
+			return nil, fmt.Errorf("expected *api.ServiceAccount, got %#v", sa)
+		}
+		// Ensure UID matches if given
+		if len(uid) == 0 || uid == sa.UID {
+			return sa, nil
+		}
+	}
+
+	if !fetchOnCacheMiss {
 		return nil, nil
 	}
 
-	key := &api.ServiceAccount{ObjectMeta: api.ObjectMeta{Namespace: secret.Namespace}}
-	namespaceAccounts, err := e.serviceAccounts.Index("namespace", key)
+	// Live lookup
+	sa, err := e.client.Core().ServiceAccounts(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		return nil, nil
+	}
 	if err != nil {
 		return nil, err
 	}
-
-	for _, obj := range namespaceAccounts {
-		serviceAccount := obj.(*api.ServiceAccount)
-
-		if serviceaccount.IsServiceAccountToken(secret, serviceAccount) {
-			return serviceAccount, nil
-		}
+	// Ensure UID matches if given
+	if len(uid) == 0 || uid == sa.UID {
+		return sa, nil
 	}
+	return nil, nil
+}
 
-	if fetchOnCacheMiss {
-		serviceAccount, err := e.client.Core().ServiceAccounts(secret.Namespace).Get(name)
-		if apierrors.IsNotFound(err) {
-			return nil, nil
+func (e *TokensController) getSecret(ns string, name string, uid types.UID, fetchOnCacheMiss bool) (*api.Secret, error) {
+	// Look up in cache
+	obj, exists, err := e.secrets.GetByKey(ns + "/" + name)
```

Comment:
hidden dependency on the metanamespacekey function. Maybe add a function to do this construction.